lib/SILOptimizer/LoopTransforms/COWArrayOpt.cpp

```diff
@@ -1446,21 +933,102 @@ bool COWArrayOpt::hasLoopOnlyDestructorSafeArrayOperations() {
   return ReturnWithCleanup(true);
 }
 
-/// Hoist the make_mutable call and optionally the projection chain that feeds
-/// the array self argument.
-void COWArrayOpt::hoistMakeMutableAndSelfProjection(
-    ArraySemanticsCall MakeMutable, bool HoistProjection) {
-  // Hoist projections.
-  if (HoistProjection)
-    hoistAddressProjections(MakeMutable.getSelfOperand(),
-      Preheader->getTerminator(), DomTree);
+/// Return the underlying Array address after stripping off all address
+/// projections. Returns an invalid SILValue if the array base does not dominate
+/// the loop.
+SILValue COWArrayOpt::getArrayAddressBase(SILValue V) {
+  while (true) {
+    V = stripSinglePredecessorArgs(V);
+    if (auto *RefCast = dyn_cast<UncheckedRefCastInst>(V)) {
+      V = RefCast->getOperand();
+      continue;
+    }
+    if (auto *SE = dyn_cast<StructExtractInst>(V)) {
+      V = SE->getOperand();
+      continue;
+    }
+    if (auto *IA = dyn_cast<IndexAddrInst>(V)) {
+      SILBasicBlock *IndexBlock = IA->getIndex()->getParentBlock();
+      if (IndexBlock && !DomTree->dominates(IndexBlock, Preheader))
```

Comment:
It's because all projections have to be loop invariant. index_addr is different from other projection instructions because it has a second operand which has to be checked separately.
I'll add a comment